### ISSUE TITLE
Bug: [no-unnecessary-condition] false positive with non-null-assertions and noUncheckedIndexedAccess

### ISSUE THREAD

**Issue by kiliancs** [typescript-eslint/typescript-eslint#7693]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.2.2&fileType=.ts&code=GYVwdgxgLglg9mABAGwIYCMCmyDy6BWm0AzgBQCwAUIomKgLabEBcixUATjGAOYDaAXSoBKRAG8qNYHA6JSyTFEQxEAXkQAGANzLEAHloMmAOgW8oACx0wA1DdETqNRBATtDjNR6Z8YArZLOHESYMABumAByRqR0jACEwgFOAL5UaZRUoJCwCIjBEKER0YyxRqzsXLwOgcFQIBxIcZjJKUA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RIAGnBgAviD1A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkRsAqkdABbrQDW6AJgJJHd0ADx4BBaNHQkylPJgCu6ADTgakTFNioAbugBSJYgFlY3eRgpg5ilTSgk5yaHkvX0qgL4gPQA&tokens=false

### Repro Code

```typescript
function labelObjects(
  names: string[]
) {
  for (let i = 0; i < names.length; i++) {
    const name = names[i];
    receiveName(name!);
  }
}

function receiveName(name: string) {
  return name;
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-condition": "error",
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "noUncheckedIndexedAccess": true,
    "resolveJsonModule": true,
    "strict": true
  }
}
```

### Expected Result

No violation of @typescript-eslint/no-unnecessary-type-assertion

### Actual Result

Violation of @typescript-eslint/no-unnecessary-type-assertion:
This assertion is unnecessary since it does not change the type of the expression. 6:17 - 6:22

### Additional Info

Somewhat related to #6632 and #7610.

**Comment by Josh-Cena** [comment]
This is because you are not asserting on an array index expression. Our rule only special-cases the pattern `array[i]!`. I'm not sure why the requested type for `name` doesn't include `undefined`, though.

**Comment by bradzacher** [comment]
> I'm not sure why the requested type for name doesn't include undefined, though.

This was what I wasn't sure about either - the type doesn't include `undefined`, yet intellisense and the type checker can see it.

**Comment by kiliancs** [comment]
Thanks for taking a look.

> This is because you are not asserting on an array index expression.

This seems less of a reason for the behavior and more of a workaround. However, I would like to be able to assign `array[i]` to a variable and then use the non-null-assertion locally. I don't want to pollute the rest of the code that already works great with `string | undefined` with a false type of `string`. For the same reason we avoid `any` but, if you had to `any` something, you'd do it as locally as possible (e.g. call site rather than variable declaration).

> the type doesn't include undefined, yet intellisense and the type checker can see it

This seems suspicious to me because it aligns with the rule thinking that the non-null assertion (`!`) doesn't make a difference (it's "unnecessary"). It is a `string | undefined`, and `!` makes it a `string` so that it can happily be passed to a function that expects a `string`. The `!` is not "unnecessary".

Not that it sheds any light, but for practical purposes here's a [smaller repro](https://typescript-eslint.io/play/#ts=5.2.2&showAST=types&fileType=.ts&code=CYUwxgNghgTiAEYD2A7AzgF3iqBbEaAXPJjAJYoDmA2gLoDcAsAFAvLpYxQDuAMlACMQEeAF5seAtQAMDNqkzxoQiMVIVKY%2BFz6DhAQnpA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RIAGnBgAviD1A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkRsAqkdABbrQDW6AJgJJHd0ADx4BBaNHQkylPJgCu6ADTgakTFNioAbugBSJYgFlY3eRgpg5ilTSgk5yaHkvX0qgL4gPQA&tokens=false).

**Comment by Josh-Cena** [comment]
Not to say there's anything working as intended here, just want to point out that your issue is completely unrelated to https://github.com/typescript-eslint/typescript-eslint/issues/6632 and https://github.com/typescript-eslint/typescript-eslint/issues/7610—those are about `array[i]!` _getting reported_ when `noUncheckedIndexedAccess` is off. The `!` is definitely not unnecessary, but somehow when we ask TS "is this type nullish" TS tells us "no".

**Comment by Josh-Cena** [comment]
Wow, TS does think this type does not include `undefined`:

<img width="663" alt="image" src="https://github.com/typescript-eslint/typescript-eslint/assets/55398995/77be21e8-60d2-420a-9045-996a7f2bc2bf">

How are we supposed to know ??

**Comment by j4k0xb** [comment]
I think this is caused by enabling `resolveJsonModule` in the playground.
If you switch to the tsconfig tab: 
![image](https://github.com/typescript-eslint/typescript-eslint/assets/55899582/6d393f90-cf87-4a9d-bbe8-ae1acaa897e6)


Removed `resolveJsonModule` - no errors: [playground](
https://typescript-eslint.io/play/#ts=5.2.2&fileType=.ts&code=CYUwxgNghgTiAEYD2A7AzgF3iqBbEaAXPJjAJYoDmA2gLoDcAsAFAvLpYxQDuAMlACMQEeAF5seAtQAMDNqkzxoQiMVIVKY%2BFz6DhAQnpA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RIAGnBgAviD1A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkRsAqkdABbrQDW6AJgJJHd0ADx4BBaNHQkylPJgCu6ADTgakEnOTQ8FMHMWqAviENA&tokens=false)
![image](https://github.com/typescript-eslint/typescript-eslint/assets/55899582/d4f955f4-306c-40b3-948f-fac9e3ecc81a)

**Comment by kiliancs** [comment]
Very interesting! Although I was running into this in my project, and the playground was just an attempt at a minimal repro, so this doesn't fully explain my case.

Is the classic module resolution necessary for the playground, given the limitations to the functionality of the playground?
